File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy
        count = 1
        while head:
            if count != left:
                curr.next = head
                head = head.next
                curr = curr.next
            else:
                prev = None
                hold = head
                while count <= right:
                    temp = head.next
                    head.next = prev
                    prev = head
                    head = temp
                    count += 1
                curr.next = prev
                while prev:
                    curr = curr.next
                    prev = prev.next
            count += 1
                
        return dummy.next
                
    
                    
        

File: 0092-reverse-linked-list-ii/0092-reverse-linked-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        int length = 1;
        ListNode* leftpt;
        ListNode* rightpt;
        ListNode* nextpt;
        ListNode* temp;
        ListNode* dummy = new ListNode();
        ListNode* curr = dummy;
        while(head){
            if(length == left){
                leftpt = head;
            }
            else if(length == right){
                rightpt = head;
                nextpt = head->next;
                break;
            }
            else if(length < left){
                curr->next = head;
                curr = curr->next;
            }
            length ++;
            head = head->next;
        }
        ListNode* prev = nullptr;
        int numREV = 0;
        while(numREV <= right - left){
            temp = leftpt->next;
            leftpt->next = prev;
            prev = leftpt;
            leftpt = temp;
            numREV ++;
        }
        curr->next = prev;
        while(curr->next){
            curr = curr->next;
        }
        curr->next = temp;
        return dummy->next;
        
        
    }
};